src/logic/map_objects/world/critter.cc

```diff
@@ -230,25 +277,192 @@ Simply roam the map
+	Time idle_time_rnd = kCritterMaxIdleTime;
 	if (state.ivar1) {
 		state.ivar1 = 0;
-		if (start_task_movepath(game,
-		                        game.random_location(get_position(), 2),  //  Pick a random target.
-		                        3, descr().get_walk_anims()))
-			return;
+		// lots of magic numbers for reasonable weighting of various nearby animals
+		uint32_t roaming_dist = 24;
+		if (game.map().find_bobs(
+		       game, Area<FCoords>(get_position(), 2), nullptr, FindBobByName(descr().name())) > 4) {
+			// Too crowded here! Let's go further away
+			roaming_dist += 3 *
+			                game.map().find_bobs(game, Area<FCoords>(get_position(), 5), nullptr,
+			                                     FindBobByName(descr().name()));
+		}
+		// move only a bit away from other animals of the same class
+		roaming_dist += 4 *
+		                game.map().find_bobs(game, Area<FCoords>(get_position(), 3), nullptr,
+		                                     FindCritterByClass(descr()));
+		if (!descr().is_carnivore()) {
+			roaming_dist +=
+			   6 *
+			   game.map().find_bobs(game, Area<FCoords>(get_position(), 4), nullptr, FindCarnivores());
+		}
+		roaming_dist /= 12;
+		assert(roaming_dist >= 2);
+		// the further we want to go, the harder we try to go somewhere
+		for (; roaming_dist; --roaming_dist) {
+			if (start_task_movepath(game, game.random_location(get_position(), roaming_dist),
+			                        roaming_dist, descr().get_walk_anims())) {
+				return;
+			}
+		}
 		idle_time_min = 1;
 		idle_time_rnd = 1000;
 	}
 	state.ivar1 = 1;
-	return start_task_idle(
-	   game, descr().get_animation("idle", this), idle_time_min + game.logic_rand() % idle_time_rnd);
+	std::vector<Critter*> candidates_for_eating;  // nullptr indicates the immovable here
+	if (descr().is_herbivore() && get_position().field->get_immovable()) {
+		for (uint32_t a : descr().food_plants()) {
+			if (get_position().field->get_immovable()->descr().has_attribute(a)) {
+				candidates_for_eating.push_back(nullptr);
+				break;
+			}
+		}
+	}
+	bool other_herbivores_on_field = false;
+	size_t mating_partners = 0;
+	int32_t n_th_bob_on_field = 0;
+	bool foundme = false;
+	for (Bob* b = get_position().field->get_first_bob(); b; b = b->get_next_bob()) {
+		assert(b);
+		if (b == this) {
+			assert(!foundme);
+			foundme = true;
+			continue;
+		}
+		if (!foundme) {
+			++n_th_bob_on_field;
+		}
+		if (descr().name() == b->descr().name()) {
+			++mating_partners;
+		}
+		if (upcast(Critter, c, b)) {
+			other_herbivores_on_field |= c->descr().is_herbivore();
+		}
+	}
+	assert(foundme);
+	if (descr().is_carnivore()) {
+		// only hunt other carnivores if there are no herbivores here
+		for (Bob* b = get_position().field->get_first_bob(); b; b = b->get_next_bob())
+			if (descr().name() != b->descr().name())
+				if (upcast(Critter, c, b))
+					if (!c->descr().is_carnivore() || !other_herbivores_on_field)
+						candidates_for_eating.push_back(c);
```

Comment:
How about doing this in the loop above st `other_herbivores_on_field |= c->descr().is_herbivore();`? Will save us a loop and an upcast.

Reply:
When deciding whether to run `candidates_for_eating.push_back` we already need to know if there are `other_herbivores_on_field`, so two loops are needed. But I got rid of the additional upcast so this should be more efficient now.

Reply:
How about using a counting variable?

Reply:
EEk. Let's leave it for now though.